rust/api/src/kv_store.rs

```diff
@@ -22,3 +25,602 @@ pub trait KvStore: Send + Sync {
 		&self, user_token: String, request: ListKeyVersionsRequest,
 	) -> Result<ListKeyVersionsResponse, VssError>;
 }
+
+#[macro_export]
+macro_rules! define_kv_store_tests {
+	($test_suite_name:ident, $store_type:path, $create_store_expr: expr) => {
+		use crate::api::error::VssError;
+		use crate::api::kv_store::KvStoreTestSuite;
+		use async_trait::async_trait;
+
+		struct $test_suite_name;
+
+		#[async_trait]
+		impl KvStoreTestSuite for $test_suite_name {
+			type Store = $store_type;
+
+			async fn create_store() -> Self::Store {
+				$create_store_expr
+			}
+		}
+
+		macro_rules! create_test {
+			($test_fn:ident) => {
+				#[tokio::test]
+				async fn $test_fn() -> Result<(), VssError> {
+					$test_suite_name::$test_fn().await?;
+					Ok(())
+				}
+			};
+		}
+
+		create_test!(put_should_succeed_when_single_object_put_operation);
+		create_test!(put_should_succeed_when_multi_object_put_operation);
+		create_test!(put_should_fail_when_key_version_mismatched);
+		create_test!(put_multi_object_should_fail_when_single_key_version_mismatched);
+		create_test!(put_should_fail_when_global_version_mismatched);
+		create_test!(put_should_succeed_when_no_global_version_is_given);
+		create_test!(put_and_delete_should_succeed_as_atomic_transaction);
+		create_test!(delete_should_succeed_when_item_exists);
+		create_test!(delete_should_succeed_when_item_does_not_exist);
+		create_test!(delete_should_be_idempotent);
+		create_test!(get_should_throw_no_such_key_exception_when_key_does_not_exist);
+		create_test!(get_should_return_correct_value_when_key_exists);
+		create_test!(list_should_return_paginated_response);
+		create_test!(list_should_honour_page_size_and_key_prefix_if_provided);
+		create_test!(list_should_return_zero_global_version_when_global_versioning_not_enabled);
+		create_test!(list_should_limit_max_page_size);
+	};
+}
+
+#[async_trait]
+pub(crate) trait KvStoreTestSuite {
+	type Store: KvStore + 'static;
+
+	async fn create_store() -> Self::Store;
+
+	async fn put_should_succeed_when_single_object_put_operation() -> Result<(), VssError> {
+		let kv_store = Self::create_store().await;
+		let ctx = TestContext::new(&kv_store);
```

Comment:
Can DRY these tests up by having each take a `&TestContext` created in `create_test!`.